xlators/mgmt/glusterd/src/glusterd-utils.c

```diff
@@ -2268,10 +2271,9 @@ glusterd_volume_start_glusterfs(glusterd_volinfo_t *volinfo,
 
     if (wait) {
         synclock_unlock(&priv->big_lock);
-        errno = 0;
         ret = runner_run(&runner);
-        if (errno != 0)
-            ret = errno;
+        if (ret)
```

Comment:
A failure running the program doesn't necessarily mean that it has failed because of any error binding the port. In fact if we reach here, the port has not hat any problem (it was bound successfully because `pmap_assign_port()` succeeded).

Reply:
This happens in a race scenario when the `pmap_assign_port()` will succeed by assigning the same ports to two different bricks (due to race conditions) and in that case when the `bind()` function executes it returns an error as `EADDRINUSE`....the situation was hit in a bz and tested using gdb.

Reply:
Can you let me know the BZ ?

In any case, you need to know the error that the executed program has got, not the possible error that was generated during `bind()` while allocating a free port.

Reply:
Sure, here it is BZ link: https://bugzilla.redhat.com/show_bug.cgi?id=1865796
The error code received was due to `bind()`, because of a race condition due to which an already occupied port was allocated to another brick.

Reply:
I don't see how this patch fixes the bug. Suppose bind succeeds, so errno is 0 (or has any other stale value, most probably not `EADDRINUSE`). Before the brick starts, another unrelated process uses the same port. The brick will fail because the port is in use, but the saved errno will be useless to detect that.

Reply:
Agreed. What I was thinking was a bit wrong, I assumed the race condition might occur in between the assignment of the port and the binding of it (though in that case, bind would have failed) while manipulating using gdb and creating a race scenario manually I got into this confusion. But, in reality, the race condition might occur between the assignment and the execution of the `runner_run()` function, which should then return us the correct `errno`

Though the previous fix was done around that and the same was being checked, but incidentally, we never got the correct `errno` from the `runner_run()` function, and the `retry` section was never hit.